ARROW-15619: [C++] Temporal component extraction function for extracting is_leap_year indicator

rok opened the pull request:
This is to resolve [ARROW-15619](https://issues.apache.org/jira/browse/ARROW-15619)

lidavidm commented on docs/source/cpp/compute.rst:
```diff
@@ -1388,6 +1388,8 @@ For timestamps inputs with non-empty timezone, localized timestamp components wi
 +--------------------+------------+-------------------+---------------+----------------------------+-------+
 | iso_calendar       | Unary      | Temporal          | Struct        |                            | \(3)  |
 +--------------------+------------+-------------------+---------------+----------------------------+-------+
+| is_leap_year       | Unary      | Temporal          | Boolean       |                            |       |
```
Maybe be explicit and list `Timestamp, Date`?

rok replied:
Indeed! Fixed.

lidavidm commented:
And now the merge conflict is here :)

rok commented:
@lidavidm rebased.

jorisvandenbossche commented:
Is there a test for a timestamp with timezone where it would give a different result if the timezone was ignored? (to catch that the timezone is properly applied) Such as "2012-01-01 00:00:00+02:00" (which is "2011-12-31 22:00:00" in the underlying UTC value, and then not a leap year)

(didn't check the exact values in the tests, so there might already be a case like this)

rok commented:
@jorisvandenbossche good point, I'm not sure either. Will add.

rok commented:
@jorisvandenbossche added `Pacific/Pago_Pago` test to show results are different than in `Australia/Broken_Hill`. Also added a Pandas test and noticed I forgot to add one when adding `is_dst` kernel, so I added one for that as well.
